crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs

```diff
@@ -0,0 +1,421 @@
+            }
+        }
+    })
+    .await
+    .unwrap();
+
+    assert_eq!(1, agents[should_recv_idx_1].0.op_store().read().len());
+
+    tokio::time::timeout(std::time::Duration::from_secs(60), {
+        let op_store_recv = agents[should_recv_idx_2].0.op_store().clone();
+        async move {
+            loop {
+                if !op_store_recv.read().is_empty() {
+                    break;
+                }
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            }
+        }
+    })
+    .await
+    .unwrap();
+
+    assert_eq!(1, agents[should_recv_idx_2].0.op_store().read().len());
+
+    let mut agents_not_receiving_count = 0;
+    for i in 0..5 {
+        if i == sender_idx || i == should_recv_idx_1 || i == should_recv_idx_2 {
+            continue;
+        }
+
+        // We've filtered out the sender and the receivers, who are expected to have the data.
+        // Now we check that the agent at the current index does not have the basis that the op was
+        // published to in its arc. That would make the test wrong, not Kitsune.
+        let should_this_agent_hold_the_op =
+            should_agent_hold_op_at_basis(&agents[i].0, agents[i].2.clone(), basis.clone());
+
+        if should_this_agent_hold_the_op {
+            tracing::warn!(
+                "Agent {} should not have received the data, this means that the test data construction isn't perfectly accurate but the test can continue",
+                i
+            );
+
+            continue;
+        }
+
+        // Now make the important assertion that the agent at index `i` did not receive the data! If it's not in the agents arc
+        // (which we just asserted above) then it should not have been received.
+        let store_lock = agents[i].0.op_store();
+        let store = store_lock.read();
+        assert!(
+            store.is_empty(),
+            "Agent {} should not have received any data but has {} ops. Ops stare: {:?}",
+            i,
+            store.len(),
+            store,
+        );
+        agents_not_receiving_count += 1;
+    }
+
+    // A slightly looser assertion than requiring the test data to be constructed perfectly with no extra overlap.
+    assert!(
+        agents_not_receiving_count >= 1,
+        "At least one agent should not have received the data"
+    );
+}
+
+fn should_agent_hold_op_at_basis(
+    kitsune_test_harness: &KitsuneTestHarness,
+    agent: Arc<KitsuneAgent>,
+    basis: Arc<KitsuneBasis>,
+) -> bool {
+    // Find the agent info for the given agent
+    let agent_store = kitsune_test_harness.agent_store();
+    let agent_store_lock = agent_store.read();
+    let agent_info = agent_store_lock
+        .iter()
+        .find(|info| info.agent == agent)
+        .unwrap();
+
+    // TODO Why is this different to doing `to_dht_arc_range_std()` and then checking contains?
+    agent_info.storage_arc().contains(&basis.get_loc())
```

Comment:
This was the problem with my tests being flaky, it was using slightly different code to check for 'arc contains' than the publish selection strategy was using. I've made them match which lets me write stable tests for the feature - I'm a little concerned that different ways of asking for the arc containing a basis can give different results. This TODO will need following up on

Reply:
We could write proptests for any multiple ways of doing things to find discrepancies, or refactor to use only one way.